cmd/ci-operator/main.go

```diff
@@ -2060,6 +2064,17 @@ func (o *options) loadConfig(info *api.Metadata) (*api.ReleaseBuildConfiguration
 	return &configSpec, nil
 }
 
+func (o *options) resolveNodeArchitectures(ctx context.Context, client coreclientset.NodeInterface) {
+	nodeList, err := client.List(ctx, meta.ListOptions{})
+	if err != nil {
+		logrus.WithError(err).Warn("failed to determine the node architectures, defaulting to amd64...")
```

Comment:
Why not return the error?  Returning `AMD64Arch` will result in an unschedulable `Build` if the cluster has no nodes with that architecture.

Reply:
> Some comments on the implementation below, but this is missing some… architectural explanation.
> 
> For instance:
> 
> * How will this cluster (these clusters?) be incorporated in our collection of build clusters?  Will it/they be responsible for _all_ image build post-submit jobs?
> * As mentioned in the issue ([DPTP-3319](https://issues.redhat.com/browse/DPTP-3319)) and in a previous discussion, using node information has the potential to miss architectures for which there are currently no nodes in the cluster.  Have we decided this is the desired behavior?  Why?
> * Does a heterogeneous cluster not have the information of which architectures it contains?  Our build clusters get fairly big when stressed, and from my experience listing nodes is a very slow (so expensive, I suppose) operation.
> * Do we want to always trust this list and not have a list of architectures for which we definitely want to produce images?

- Once we start introducing the heterogenous clusters, I would expect test and image promotions should work the same way but with the multiple manifest lists included. We need to make sure that we build all the images for all architectures on tests as well because otherwise, we will have to nodeSelect every pod. If the architecture is available in the image, then the pod will pull the correct manifest despite the node that it's running. We will start by scheduling the postsubmits in the heterogenous clusters to build the image for all available architectures. (This requires app.ci to be 4.13 version)
- I had a discussion with the cloud team and it seems that there is no way to get the architecture from a machine set. The only case we need to cover is that if a machineset is scaled down to 0 but the implementation is still in progress. They suggested manually maintaining the machineset and applying a specific label that we maintain. I would prefer to start by gathering the available architectures from the active nodes, and as a follow-up we can cover the scaled-down-to-0 machinesets design...
- Nope, there is no way to get the architectures. I tried to find it but these kind of clusters are work-in-progress too.
- I would say to trust this list because it will still work once we add more architectures in any of the clusters.

Reply:
That takes care of my initial questions, thanks.